### PR TITLE
Add power capping API for NVIDIA, AMD and Intel GPUs

### LINKED ISSUE #316: Add GPU power capping capabilities
For NVIDIA, AMD and Intel DGPU ports.

### PR DESCRIPTION
**Pull request by amarathe84**
# Description

This is a PR to add Variorum API to enable GPU power capping for NVIDIA, AMD and Intel GPUs. This PR fixes #316. 

## Type of change

- New feature/architecture support: Cap power usage of NVIDIA, AMD and Intel GPUs wherever supported.
  - [x] Nvidia
  - [x] AMD
  - [x] Intel
- [x] Documentation update
- [x] Build/CI update

# Testing: 

The contributions in this PR have been successfully tested on the following GPU platforms:

- NVIDIA: Volta GPUs on Lassen and Alehouse (internal system)
- AMD: Instinct MI-100 on Chameleon Cluster  (www.chameleoncloud.org)
- Intel GPU: Intel DGPU system at Argonne National Laboratory. 

----
Log:
- Continuation of old PR #365 with updated name for the originating branch

### CODE REVIEW

File: host-configs/chameleon-epyc-x86_64-gcc@9.4.0-rocm@5.2.cmake [hunk 1]
```diff
@@ -0,0 +1,39 @@
+# Copyright 2019-2022 Lawrence Livermore National Security, LLC and other
```
Comment (slabasan) [inline host-configs/chameleon-epyc-x86_64-gcc@9.4.0-rocm@5.2.cmake:1]: Update copyright year to 2023 on new files

File: src/examples/variorum-cap-gpu-power-limit-example.c [hunk 2]
```diff
@@ -0,0 +1,60 @@
+// Copyright 2019-2022 Lawrence Livermore National Security, LLC and other
```
Comment (slabasan) [inline src/examples/variorum-cap-gpu-power-limit-example.c:1]: Update copyright to 2023

File: src/variorum/AMD_GPU/config_amd_gpu.c [hunk 3]
```diff
@@ -33,6 +33,10 @@ int set_amd_gpu_func_ptrs(int idx)
         g_platform[idx].variorum_print_power_limit     = instinctGPU_get_power_limit;
         g_platform[idx].variorum_print_gpu_utilization =
             instinctGPU_get_gpu_utilization;
+
```
Comment (slabasan) [inline src/variorum/AMD_GPU/config_amd_gpu.c]: Remove extra line

File: src/variorum/Intel_GPU/GPU.c [hunk 4]
```diff
@@ -56,3 +56,38 @@ int intel_gpu_get_clocks(int long_ver)
...
+    unsigned nsockets;
+    variorum_get_topology(&nsockets, NULL, NULL, P_INTEL_GPU_IDX);
+    for (iter = 0; iter < nsockets; iter++)
+    {
+        get_power_limit_data(iter, long_ver, stdout);
+    }
+    return 0;
+}
+
+
```
Comment (slabasan) [inline src/variorum/Intel_GPU/GPU.c]: Extra lines can be removed

File: src/variorum/Nvidia_GPU/config_nvidia.c [hunk 5]
```diff
@@ -30,6 +30,9 @@ int set_nvidia_func_ptrs(int idx)
         g_platform[idx].variorum_print_frequency       = volta_get_clocks;
         g_platform[idx].variorum_print_power_limit     = volta_get_power_limits;
         g_platform[idx].variorum_print_gpu_utilization = volta_get_gpu_utilization;
+
```
Comment (slabasan) [inline src/variorum/Nvidia_GPU/config_nvidia.c]: Remove extra line

File: src/variorum/variorum.h [hunk 6]
```diff
@@ -149,6 +142,18 @@ int variorum_cap_each_core_frequency_limit(int cpu_freq_mhz);
 /// not supported, otherwise -1
 int variorum_cap_socket_frequency_limit(int socketid, int socket_freq_mhz);
 
+/// @brief Cap the power usage identically of each GPU on the node.
+///
+/// @supparch
+/// - NVIDIA Volta and later
+/// - WIP: AMD
```
Comment (slabasan) [inline src/variorum/variorum.h]: Verify supported architectures, add Intel Discrete GPU

File: src/docs/sphinx/IntelGPU.rst [hunk 7]
```diff
@@ -103,6 +103,22 @@ Variorum provides an API to report instantaneous clock speed of the Intel GPU's
...
+the corresponding APMIDG APIs.
+
+Power control
+=============
+
+Variorum provides an API to cap the power usage of an Intel Discrete GPU device.
+To set the power cap, Variorum uses the ``apmidg_setpwrlim()`` API of APMIDG
+which takes as input the GPU device ID, the power domain ID and the power cap in
+milliwatts.
+
```
Comment (tpatki) [inline src/docs/sphinx/IntelGPU.rst:120]: Very minor comment, feel free to ignore. 

The text in this rst file seems a little repetitive: 
In the the "Power control" subsection, you can maybe remove the first sentence ("Variorum provides...device"), and edit the second sentence to say "In Variorum's GPU power capping API, Variorum uses the ``apmidg....`` API, which takes as input...milliwatts".

File: src/variorum/Intel_GPU/config_intel_gpu.c [hunk 8]
```diff
@@ -26,7 +26,8 @@ int set_intel_gpu_func_ptrs(int idx)
         g_platform[idx].variorum_print_power           = intel_gpu_get_power;
         g_platform[idx].variorum_print_thermals        = intel_gpu_get_thermals;
         g_platform[idx].variorum_print_frequency       = intel_gpu_get_clocks;
-        // g_platform.variorum_print_power_limit    = ; // implement this later
+        g_platform[idx].variorum_print_power_limit     = intel_gpu_get_power_limit;
```
Comment (tpatki) [inline src/variorum/Intel_GPU/config_intel_gpu.c]: In addition to the new API for cap_gpu_power limit, this PR also implements the print power limit for Intel DGPU (not a new API, but missing functionality on Intel DGPU) -- can you update the docs to reflect this?
My guess is @slabasan would want this to be a separate PR, I am ok either way, as in, we can merge it in with this PR.

File: src/variorum/variorum.h [hunk 9]
```diff
@@ -149,6 +142,18 @@ int variorum_cap_each_core_frequency_limit(int cpu_freq_mhz);
...
+///
+/// @param [in] gpu_power_limit Desired power limit in watts for each GPU
+///             on the node.
+///
+/// @return 0 if successful, otherwise -1
+int variorum_cap_gpu_power_limit(int gpu_power_limit);
```
Comment (tpatki) [inline src/variorum/variorum.h]: Do we want to call this `variorum_cap_gpu_power_limit` or `variorum_cap_each_gpu_power_limit`? Our convention so far was to use `each` to indicate a uniform power cap on APIs where we don't specify the device ID. (See the socket power cap or frequency cap examples). We are going to redesign the power capping API soon, so not sure if this matters. 

But in your PR, you deleted the previously commented version of `cap_each_gpu_power_limit` and dropped the `each` -- not sure if we want to keep it. Will defer to Stephanie here.  @slabasan Please comment.

In general, in the future, we want to say something clearer in the APIs: eg `cap_uniform_power_limit(device type, device ID, power limit)` for example (divide type can be socket, core, Memory, GPU, device ID is the numeric ID of the device, eg the 10th core or 3rd GPU, and the limit is self-explanatory).

File: src/variorum/Intel_GPU/power_features.c [hunk 10]
```diff
@@ -134,3 +134,63 @@ void get_clocks_data(int chipid, int verbose, FILE *output)
...
+                        "_INTEL_GPU_POWER_LIMIT Host Socket DeviceID GPU_Power_limit_mW\n");
+                init_output = 1;
+            }
+            fprintf(output, "_INTEL_GPU_POWER_LIMIT %s %d %d %d\n",
+                    m_hostname, chipid, d, current_powerlimit_mwatts);
+        }
+    }
+}
+
+
```
Comment (slabasan) [inline src/variorum/Intel_GPU/power_features.c]: Just need 1 new line between functions

File: src/variorum/Nvidia_GPU/power_features.c [hunk 11]
```diff
@@ -212,3 +212,24 @@ void get_gpu_utilization(int chipid, int verbose, FILE *output)
...
+        if (NVML_SUCCESS != nvmlDeviceSetPowerManagementLimit(
+                m_unit_devices_file_desc[d], powerlimit_mwatts))
+        {
+            variorum_error_handler("Could not set the specified GPU power limit",
+                                   VARIORUM_ERROR_PLATFORM_ENV, getenv("HOSTNAME"), __FILE__, __FUNCTION__,
+                                   __LINE__);
+        }
+    }
+}
+
```
Comment (slabasan) [inline src/variorum/Nvidia_GPU/power_features.c]: Remove new line

File: src/variorum/variorum.h (hunk 9 above)
Comment (slabasan) [reply src/variorum/variorum.h]: I think we want to use `each` if we've been using that in existing convention. Then we can make the change to something more clear and specific across the board at once.

File: src/variorum/Intel_GPU/config_intel_gpu.c (hunk 8 above)
Comment (slabasan) [reply src/variorum/Intel_GPU/config_intel_gpu.c]: Ah I didn't catch that, yes would prefer if that went into another PR. How much of the print power limit for Intel DGPU implementation is in this PR? Docs and implementation? Just implementation?

Comment (tpatki) [reply src/variorum/Intel_GPU/config_intel_gpu.c]: I just saw the implementation, I didn't see the associated docs for this in this PR. I think removing from this one and adding that in another PR is a good idea. Tagging @amarathe84 so he sees it

File: src/variorum/variorum.h (hunk 9 above)
Comment (tpatki) [reply src/variorum/variorum.h]: Agreed, so then in this PR, we need to change `variorum_cap_gpu_power_limit` --> `variorum_cap_each_gpu_power_limit` everywhere. Tagging @amarathe84 so he gets notified.

Comment (tpatki) [reply src/variorum/variorum.h]: Apparently I need to leave a comment for me to indicate "changes requested" in the review, so here it is :D

**Review by tpatki — CHANGES_REQUESTED** [review]

File: src/variorum/variorum.h (hunk 9 above)
Comment (amarathe84) [reply src/variorum/variorum.h]: I think I was following the API naming convention of `variorum_cap_gpu_power_ratio` and `variorum_cap_socket_frequency_limit` that's defined just above the new API. Both these APIs don't have 'each' but apply to each socker or GPU internally, so I had previously skipped 'each' in the API name. 

I agree with the suggested change to `variorum_cap_each_gpu_power_limit` as it makes more sense (also updating the docs). We'll streamline the other two APIs noted above in a separate PR for API rehaul work.

**Review by slabasan — APPROVED** [review]
